src/commonMain/kotlin/org/kosat/heuristics/Preprocessor.kt

```diff
@@ -0,0 +1,279 @@
+            isClauseDeleted[clause] = true
+        }
+    }
+
+    // remove subsumed clauses
+    private fun removeSubsumedClauses() {
+        val uselessClauses = mutableSetOf<Int>()
+        val duplicateClauses = mutableSetOf<Int>()
+        val markedClauses = MutableList(solver.constraints.size) { false }
+
+        // going from the end because smaller clauses appear after big one
+        for (ind in solver.constraints.lastIndex downTo 0) {
+            if (!markedClauses[ind]) {
+                findSubsumed(ind).forEach {
+                    if (!markedClauses[it]) {
+                        if (ind.clauseSize() < it.clauseSize()) {
+                            markedClauses[it] = true
+                            uselessClauses.add(it)
+                        } else if (ind.clauseSize() == it.clauseSize()) {
+                            duplicateClauses.add(it)
+                        }
+                    }
+                }
+            }
+        }
+
+        val copiedClauses = duplicateClauses.map { solver.constraints[it] }
+
+        solver.constraints.removeAll(uselessClauses.map { solver.constraints[it] })
+        // remove duplicate clauses and leave 1 copy of each
+        solver.constraints.removeAll(copiedClauses)
+        solver.constraints.addAll(copiedClauses)
+        countOccurrence()
+        updateSig()
+    }
+
+    private fun removeTautologies() {
+        val isClauseRemoved = MutableList(solver.constraints.size) { false }
+        for (lit in 1..solver.varsNumber) {
+            val sz1 = litOccurrence[litPos(lit)].size
+            val sz2 = litOccurrence[litPos(-lit)].size
+            if (sz1 == 0 || sz2 == 0) {
+                continue
+            }
+            var ind1 = 0
+            var ind2 = 0
+            while (ind2 < sz2) {
+                while (ind1 < sz1 && litOccurrence[litPos(lit)][ind1] < litOccurrence[litPos(-lit)][ind2]) {
+                    ind1++
+                }
+                if (ind1 == sz1) {
+                    break
+                }
+                if (litOccurrence[litPos(lit)][ind1] == litOccurrence[litPos(-lit)][ind2]) {
+                    isClauseRemoved[litOccurrence[litPos(lit)][ind1]] = true
+                }
+                ind2++
+            }
+        }
+        val deletedClauses = mutableListOf<Int>()
+        isClauseRemoved.forEachIndexed { ind, isDeleted ->
+            if (isDeleted) {
+                deletedClauses.add(ind)
+            }
+        }
+        solver.constraints.removeAll(deletedClauses.map { solver.constraints[it] })
+
+        countOccurrence()
+        updateSig()
+    }
+
+    // return position of literal in occurrence array
+    private fun litPos(lit: Int): Int {
+        return if (lit >= 0) {
+            lit
+        } else {
+            solver.varsNumber - lit
+        }
+    }
+
+    private fun countOccurrence() {
+        litOccurrence = mutableListOf()
+        //litOccurrence.clear()
```

Comment:
Resolve the TODO.
There is no big difference between re-creating the list and clearing it. When you re-create, you obtain: +1 allocation (in JVM it is ultra-fast, literally 1 tick), +1 garbage (old list) (this is negligible). And when you clear: you leave the old capacity until the next realloc (since this is NOT a hotspot, you can ignore it). So, both variants are equivalent for you.